tests/basic/distribute/directory-rebalance.t

```diff
@@ -0,0 +1,114 @@
+TEST $CLI volume start $V0
+TEST $GFS --volfile-id=/$V0 --volfile-server=$H0 $M0
+TEST mkdir $M0/${DIR1}
+for i in {1..10}
+do
+    TEST_IN_LOOP dd if=/dev/zero of=$M0/$i bs=1k count=1
+    TEST_IN_LOOP dd if=/dev/zero of=$M0/${DIR1}/$i bs=1k count=1
+    TEST_IN_LOOP mknod $M0/block-${i} b 4 5
+    TEST_IN_LOOP "touch $M0/${i}\*\)\(\'" #Files with special chars in them to check that the index is built correctly and files are migrated correctly
+done
+TEST $CLI volume add-brick $V0 $H0:$B0/${V0}1 force
+TEST $CLI volume rebalance $V0 fix-layout start
+EXPECT_WITHIN $REBALANCE_TIMEOUT "fix-layout completed" fix-layout_status_field $V0
+TEST $(dirname $0)/../../../extras/rebalance/directory-rebalance.py $M0/${DIR1}
+
+root_file_success_count=0
+root_file_failure_count=0
+dir_file_success_count=0
+dir_file_failure_count=0
+for i in {1..10}
+do
+    #root directory files shouldn't be migrated
+    if [ -f $B0/${V0}0/$i ]
+    then
+        root_file_success_count=$((root_file_success_count+1))
+    fi
+
+    #Some files in ${DIR1} should be migrated
+    if [ -f $B0/${V0}0/${DIR1}/$i ]
+    then
+        dir_file_success_count=$((dir_file_success_count+1))
+    else
+        dir_file_failure_count=$((dir_file_failure_count+1))
+        if [ -f $B0/${V0}1/${DIR1}/$i ]
+        then
+            dir_file_success_count=$((dir_file_success_count+1))
+        fi
+    fi
+done
+
+EXPECT "^10$" echo $root_file_success_count
+EXPECT "^10$" echo $dir_file_success_count
+EXPECT_NOT "^0$" echo $dir_file_failure_count
+TEST $(dirname $0)/../../../extras/rebalance/directory-rebalance.py $M0
+
+root_file_success_count=0
+root_file_failure_count=0
+root_block_success_count=0
+root_block_failure_count=0
+for i in {1..10}
+do
+    if [ -e $B0/${V0}0/block-${i} ]
+    then
+        root_block_success_count=$((root_block_success_count+1))
+    else
+        root_block_failure_count=$((root_block_failure_count+1))
+        if [ -e $B0/${V0}1/block-${i} ]
+        then
+            root_block_success_count=$((root_block_success_count+1))
+        fi
+    fi
+
+    #Some files in root should be migrated
+    if [ -f $B0/${V0}0/$i ]
+    then
+        root_file_success_count=$((root_file_success_count+1))
+    else
+        root_file_failure_count=$((root_file_failure_count+1))
+        if [ -f $B0/${V0}1/$i ]
+        then
+            root_file_success_count=$((root_file_success_count+1))
+        fi
+    fi
+done
+
+EXPECT "^10$" echo $root_file_success_count
+EXPECT_NOT "^0$" echo $root_file_failure_count
+#special files also should be migrated
+EXPECT "^10$" echo $root_block_success_count
+EXPECT_NOT "^0$" echo $root_blocks_failure_count
+
+#If we trigger the rebalance command now, nothing should be rebalanced
+#Test that rebalance crawl is equivalent to manual rebalance
+TEST $CLI volume rebalance $V0 start
+EXPECT_WITHIN $REBALANCE_TIMEOUT "completed" rebalance_status_field $V0
+EXPECT "^0$" rebalanced_files_field $V0
+
```

Comment:
We have tested what is expected but do we also need to test what should not be expected.
I mean heal info? is there any empty file (.........T)? We do not have any duplicate file?
I mean the total number of files in a volume should be same as before rebalance.

Reply:
This is pure distribute volume, so heal info is not needed. If there is any difference between script execution and rebalance execution rebalanced_files_filed should have been non-zero. I can also add rebalanced_size_filed and it should be zero as well. At the end I will do umount and mount again to make sure the checksum on the data files match.